src/main/java/com/playmonumenta/networkrelay/RemotePlayerAbstraction.java

```diff
@@ -0,0 +1,128 @@
+	protected final @Nullable Boolean mIsHidden;
+	// Data provided by other plugins
+	protected final Map<String, JsonObject> mPluginData;
+
+	// Data from this (local) server
+	protected RemotePlayerAbstraction(String serverId, UUID uuid, String name, boolean isOnline, @Nullable Boolean isHidden) {
+		mServerId = serverId;
+		mUuid = uuid;
+		mName = name;
+		mIsOnline = isOnline;
+		mIsHidden = isHidden;
+		mPluginData = new HashMap<>();
+	}
+
+	// Data from a remote server of any type (subclasses provide extra information)
+	public RemotePlayerAbstraction(JsonObject remoteData) {
+		mServerId = remoteData.get("serverId").getAsString();
+		mUuid = UUID.fromString(remoteData.get("playerUuid").getAsString());
+		mName = remoteData.get("playerName").getAsString();
+		mIsOnline = remoteData.get("isOnline").getAsBoolean();
+		JsonPrimitive isHiddenJson = remoteData.getAsJsonPrimitive("isHidden");
+		if (isHiddenJson == null || !isHiddenJson.isBoolean()) {
+			mIsHidden = null;
+		} else {
+			mIsHidden = isHiddenJson.getAsBoolean();
+		}
+		mPluginData = new HashMap<>();
+		JsonObject pluginData = remoteData.getAsJsonObject("pluginData");
+		for (String key : pluginData.keySet()) {
+			mPluginData.put(key, pluginData.getAsJsonObject(key));
+		}
+	}
+
+	// Determine the appropriate remote player data type to use
+	public static RemotePlayerAbstraction from(JsonObject remoteData) {
+		String serverType = remoteData.get("serverType").getAsString();
+		switch (serverType) {
+			case RemotePlayerPaper.SERVER_TYPE:
+				return new RemotePlayerPaper(remoteData);
+			case RemotePlayerBungee.SERVER_TYPE:
+				return new RemotePlayerBungee(remoteData);
+			default:
+				return new RemotePlayerGeneric(remoteData);
+		}
+	}
+
+	// Serializes player data to be broadcast to remote servers
+	public JsonObject toJson() {
+		JsonObject playerData = new JsonObject();
+		playerData.addProperty("serverType", getServerType());
+		playerData.addProperty("serverId", mServerId);
+		playerData.addProperty("playerUuid", mUuid.toString());
+		playerData.addProperty("playerName", mName);
+		if (mIsHidden != null) {
+			playerData.addProperty("isHidden", mIsHidden);
+		}
+		playerData.addProperty("isOnline", mIsOnline);
+		playerData.add("pluginData", serializePluginData());
+		return playerData;
+	}
+
+	// Get a given plugin's data, if available
+	@Nullable
+	public JsonObject getPluginData(String pluginId) {
+		return mPluginData.get(pluginId);
+	}
+
+	protected JsonObject serializePluginData() {
+		JsonObject pluginData = new JsonObject();
+		for (Map.Entry<String, JsonObject> entry : mPluginData.entrySet()) {
+			pluginData.add(entry.getKey(), entry.getValue());
+		}
+		return pluginData;
+	}
+
+	public abstract String getServerType();
+
+	public String getServerId() {
+		return mServerId;
+	}
+
+	public UUID getUuid() {
+		return mUuid;
+	}
+
+	public String getName() {
+		return mName;
+	}
+
+	protected void broadcast() {
+		JsonObject playerData = this.toJson();
```

Comment:
`this.` is typically not required and so we usually skip it; was it using the abstraction's version of `toJson()` instead of the specific type's version? If so, a comment may prove helpful for anyone looking at this later.

Reply:
It isn't I can revert it
It was because I was investigating the mTargetShard value being null but it was just bungee being dumb.